test/types_test.py

```diff
@@ -56,3 +56,53 @@ def test_range_errors(self, n_points=1000, n_trees=10):
             self.assertRaises(IndexError, i.get_distance, 0, bad_index)
             self.assertRaises(IndexError, i.get_nns_by_item, bad_index, 1)
             self.assertRaises(IndexError, i.get_item_vector, bad_index)
+
+    def test_missing_len(self):
+        """
+        We should get a helpful error message if our vector doesn't have a
+        __len__ method.
+        """
+        class FakeCollection:
+            pass
+        
+        i = AnnoyIndex(10, 'euclidean')
+        with self.assertRaises(TypeError, msg="object of type 'FakeCollection' has no len()"):
+            i.add_item(1, FakeCollection())
+
+    def test_missing_getitem(self):
+        """
+        We should get a helpful error message if our vector doesn't have a
+        __getitem__ method.
+        """
+        class FakeCollection:
+            def __len__(self):
+                return 5
+        
+        i = AnnoyIndex(5, 'euclidean')
+        with self.assertRaises(TypeError, msg="'FakeCollection' object is not subscriptable"):
+            i.add_item(1, FakeCollection())
+
+    def test_short(self):
+        """
+        Ensure we handle our vector not being long enough.
+        """
+        class FakeCollection:
+            def __len__(self):
+                return 3
+            
+            def __getitem__(self, i):
+                raise IndexError 
+        
+        i = AnnoyIndex(3, 'euclidean')
+        with self.assertRaises(IndexError):
```

Comment:
looks like this raises an AttributeError on python 2.7, is it easy to fix? otherwise i'm maybe fine dropping support for 2.7 soon anyway...